docs/en/Unit-Testing.md

```diff
@@ -0,0 +1,78 @@
+	    }
+
+	    @After
+	    public void after() throws Exception {
+	      mMaster.stop();
+	    }
+
+2. Decide on an element of functionality to test. The functionality you decide to test should be part of the public API and should not care about implementation details.
+3. Give your test a name that describes what functionality it's testing. The functionality being tested should ideally be simple enough to fit into a name, e.g. `removeNonexistentBlockThrowsException`, `mkdirCreatesDirectory`, or `cannotMkdirExistingFile`.
+
+	    @Test
+	    public void detectLostWorker() throws Exception {
+
+4. Set up the situation you want to test. Here we register a worker and then simulate an hour passing. The `HeartbeatScheduler` section enforces that the lost worker heartbeat runs at least once.
+
+	      // Register a worker.
+	      long worker1 = mMaster.getWorkerId(NET_ADDRESS_1);
+	      mMaster.workerRegister(worker1,
+	          ImmutableList.of("MEM"),
+	          ImmutableMap.of("MEM", 100L),
+	          ImmutableMap.of("MEM", 10L),
+	          NO_BLOCKS_ON_TIERS);
+
+	      // Advance the block master's clock by an hour so that the worker appears lost.
+	      mClock.setTimeMs(System.currentTimeMillis() + Constants.HOUR_MS);
+
+	      // Run the lost worker detector.
+	      HeartbeatScheduler.await(HeartbeatContext.MASTER_LOST_WORKER_DETECTION, 1, TimeUnit.SECONDS);
+	      HeartbeatScheduler.schedule(HeartbeatContext.MASTER_LOST_WORKER_DETECTION);
+	      HeartbeatScheduler.await(HeartbeatContext.MASTER_LOST_WORKER_DETECTION, 1, TimeUnit.SECONDS);
+
+5. Check that the class behaved correctly:
+
+	      // Make sure the worker is detected as lost.
+	      Set<WorkerInfo> info = mMaster.getLostWorkersInfo();
+	      Assert.assertEquals(worker1, Iterables.getOnlyElement(info).getId());
+	    }
+6. Loop back to step #2 until the class's entire public API has been tested.
+
+# Conventions
+1. The tests for `src/main/java/ClassName.java` should go in `src/test/java/ClassNameTest.java`
+2. Tests do not need to handle or document specific checked exceptions. Prefer to simply add `throws Exception` to the test method signature.
+3. Aim to keep tests short and simple enough that they don't require comments to understand.
+
+# Patterns to avoid
+
+1. Avoid randomness. Edge cases should be handled explicitly.
+2. Avoid waiting for something by calling Thread.sleep(). This leads to slower unit tests and can cause flaky failures if the sleep isn't long enough.
+3. Avoid modifying global state. We typically use JUnit `Rule` when this is necessary so that the state is properly restored. See `SystemPropertyRule`, `TtlIntervalRule`, or `LocalAlluxioClusterResource` for examples of this.
+4. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
```

Comment:
Our codebase is still too full of singletons and final classes which don't implement interfacess